README.md
# AIA Calendar Manager - Accept Invites Automatically

## Download

<https://www.curseforge.com/wow/addons/aia>


## About

**AIA** is a simple addon to automatically accept in-game calendar invites. If your guild uses the in-game calendar for events, but you want to automatically reply that you'll be there. This is for you - especially if your guild creates events for every raid etc.



By default, AIA will ignore events where you have manually set them to tentative or declined. Using AIA, you don't have to worry about if you're signed up for an event, and if you can't attend - simply set yourself as such!


## Options

AIA comes with a few options which can be accessed from the Blizzard Interface window, on the AddOns tab, or by typing **/aia**.

* **Disable After Running** - This will shut down AIA after it has run once upon logging in. This means you don't need to worry about it using resources in the background. However, it also means if a new event is added while you are playing, AIA won't detect it until you enable AIA again, or you relog.
* **Notify when finished** - AIA will tell you how many events it has accepted you for when it has finishd its check.
* **Notify when disabled** - AIA will tell you when the addon disables itself after running.
* **Declined** - If selected, AIA will also set any events you have declined to accepted. Obviously this is disabled by default.
* **Tentative** - Same as Declined, but for Tentative status.
* **Invited** - This is the default status when you get invited to an event. If you want AIA to automatically accept any new events, leave this enabled.
* **Only Accept Invites Sent By** - Here you can enter a name or mulitple names separated by commas. If the person that created the event doesn't match what you have entered here, it will be ignored.
* **Only Accept invites with title** - Here you can enter the name of an event you want AIA to accept. You can add multiple events by separating each one with a comma. If an event title doesn't match what you have entered, it is ignored.
* **Sign Up** - If you have a guild event available for you to sign up to, but you haven't been invited to, you can select this option to have AIA automatically sign you up for it.
    * **Sign up as Tentative** - You can select this to make these Sign Ups sign up as Tentative rather than as Signed Up.
    * You can also filter these types of events by the creator name, and event title too, separately to invited events.
* **Only Accept invites of these types** - Allows you to filter invites by content type. If your guild lists meetings in the calendar as well as raids, you can choose if AIA will accept these types of events too.
* **Check Again** - If you have changed settings and think AIA might catch some other events to accept you for, you can click this, and AIA will check again. You can also type /aia check to run this same command.


## Feedback

You can [report issues on Curseforge](https://wow.curseforge.com/projects/aia/issues) or contact me directly through my [Battle.net community](https://blizzard.com/invite/WqRG7EUgOR).

## Localisation

AIA supports localisation, if you want to help me out by localising AIA, please go to the [localisation page on Curseforge](https://wow.curseforge.com/projects/aia/localization).


CHANGELOG.md
# AIA r25-Release
### Updates
* Updated interface version for 9.0.2

Core.lua
local AIA = LibStub('AceAddon-3.0'):GetAddon('AIA')
local L = LibStub('AceLocale-3.0'):GetLocale('AIA')
local LDB = LibStub('LibDataBroker-1.1'):GetDataObjectByName('AIA')
local invitesToAccept = {}
local invitesAccepted = 0
local currentDate
local currentMonth = tonumber(date('%m'))
local currentYear = tonumber(date('%Y'))
local enteredWorld = false
local firstRun = true
local elapsed = 0
local totalEvents = {}
local eventCount = 0
local sessionAccepted = 0
local currentlyChecking = false
local calendarOpen = false


function AIA:InviteStatus(event)
	if not event then return end
	if event.calendarType ~= 'PLAYER' and event.calendarType ~= 'GUILD_EVENT' and event.calendarType ~= 'COMMUNITY_EVENT' then return end

	local status = event.inviteStatus
	if not status then return end
	if status == 2 or (status >= 4 and status <= 7) then return 'Replied' end -- AIA does nothing for these statuses because they're already forms of accepted.
	-- Statuses: Invited = 1, Accepted = 2, Declined = 3, Confirmed = 4, Out = 5, Standby = 6, Signed Up = 7, Not Signed Up = 8, Tentative = 9

	local inviter = AIA:FindEventCreator(event)
	if not inviter then return end

	if event.inviteType == 1 then -- If the event is invite only or sign up. 1 == Invite, 2 == Sign Up
		if AIA:StringFilter(AIA.db.profile.Filter.Name, inviter) == false then return false end
		if AIA:StringFilter(AIA.db.profile.Filter.Title, event.title) == false then return false end
		if status == 1 then
			if AIA.db.profile.Types.Invited == true then
				return true
			end
		elseif status == 3 then
			if AIA.db.profile.Types.Declined == true then
				return true
			end
		elseif status == 9 then
			if AIA.db.profile.Types.Tentative == true then
				return true
			end
		end
	elseif event.inviteType == 2 then
		if AIA:StringFilter(AIA.db.profile.Filter.SignUp.Name, inviter) == false then return false end
		if AIA:StringFilter(AIA.db.profile.Filter.SignUp.Title, event.title) == false then return false end
		if status == 8 then
			if AIA.db.profile.Types.SignUp == true then
				return true
			end
		end
	end

	return false
end

function AIA:CreateCalendarList(eventName)
	if CalendarFrame then
		if CalendarFrame:IsVisible() == true then
			return
		end
	end
	if eventName == 'PLAYER_ENTERING_WORLD' then
		-- Force the calendar to update so we can grab data immediately.
		C_Calendar.OpenCalendar()
		enteredWorld = true
	end
	if enteredWorld == false then return end
	if currentlyChecking then return end

	wipe(invitesToAccept)
	currentDate = tonumber(date('%Y%m%d%H%M'))

	local selectedYear = currentYear
	local selectedMonth = currentMonth

	for m = 0,12 do
		currentlyChecking = true
		selectedMonth = currentMonth + m
		if selectedMonth > 12 then
			selectedMonth = selectedMonth - 12
			selectedYear = currentYear + 1
		end
		C_Calendar.SetAbsMonth(selectedMonth,selectedYear)
		for d = 1,AIA:DaysInMonth(selectedMonth,selectedYear) do
			local numDayEvents = C_Calendar.GetNumDayEvents(0,d) -- Month offset is 0 because we check set the current month above, and check only this month.
			for i = 1,numDayEvents do
				local event = C_Calendar.GetDayEvent(0,d,i) or nil
				if event then
					if event.calendarType == 'PLAYER' or event.calendarType == 'GUILD_EVENT' or event.calendarType == 'COMMUNITY_EVENT' then
						if AIA:DateConversion(event.startTime) > currentDate and not event.isLocked then
							totalEvents[tostring(0)..tostring(d)..tostring(i)] = true
							if AIA:InviteStatus(event) == 'Replied' then
								totalEvents[tostring(0)..tostring(d)..tostring(i)] = false
							end
							if AIA:EventFilter(event) == true and AIA:InviteStatus(event) == true then
								totalEvents[tostring(0)..tostring(d)..tostring(i)] = false
								local data = {
									month = selectedMonth,
									year = selectedYear,
									day = d,
									index = i,
								}
								table.insert(invitesToAccept,data)
								if #invitesToAccept > invitesAccepted then
									invitesAccepted = #invitesToAccept
								end
							end
						end
					end
				end
			end
		end
	end

	currentlyChecking = false
	AIA.Accepter:SetScript('OnUpdate', AIA.AcceptInvite)
end

function AIA:AcceptInvite(timer)
	if timer then
		if firstRun then -- Delay first run when logging in a little bit.
			elapsed = elapsed + timer
			if elapsed > 15 then
				elapsed = 0
				firstRun = false
			else return
			end
		elseif timer then
			elapsed = elapsed + timer
			if elapsed > 0.05 then
				elapsed = 0
			else return
			end
		end
	end
	if firstRun then return end
	if C_Calendar.IsActionPending() then return end

	if not invitesToAccept[1] then
		eventCount = AIA:CheckFilteredEventCount(totalEvents)

		if AIA.db.profile.NotifyFinish then
			if (AIA.db.profile.NotifyOnlyAccepted and invitesAccepted > 0) or not AIA.db.profile.NotifyOnlyAccepted then
				if invitesAccepted == 1 then
					print('|cFFFF2C5AAIA: |r'..L['Accepted ']..invitesAccepted..L[' calendar invite.'])
				else
					print('|cFFFF2C5AAIA: |r'..L['Accepted ']..invitesAccepted..L[' calendar invites.'])
				end
			end
		end

		if AIA.db.profile.WarnIgnoredEvents == true then
			if eventCount == 1 then
				print('|cFFFF2C5AAIA: |r'..eventCount..L[' calendar invite has been ignored by AIA due to your filters.'])
			elseif eventCount > 0 then
				print('|cFFFF2C5AAIA: |r'..eventCount..L[' calendar invites have been ignored by AIA due to your filters.'])
			end
		end

		if AIA.db.profile.LDB.WarnIgnoredEvents == true then
			if eventCount == 1 then
				LDB.text = string.format('|cFFFF2C5A%d|r |cFFFFFFFF'..L['Pending invite']..'|r', eventCount)
			elseif eventCount > 0 then
				LDB.text = string.format('|cFFFF2C5A%d|r |cFFFFFFFF'..L['Pending invites']..'|r', eventCount)
			else
				LDB.text = AIA.db.profile.LDB.DisplayName
			end
		else
			LDB.text = AIA.db.profile.LDB.DisplayName
		end

		if AIA.db.profile.DisableAfterRunning == true then
			AIA:Disable()
		else
			AIA.Accepter:SetScript('OnUpdate', nil)
		end

		sessionAccepted = sessionAccepted + invitesAccepted
		invitesAccepted = 0
		currentlyChecking = false

		return
	end

	currentlyChecking = true
	local d = invitesToAccept[1].day
	local i = invitesToAccept[1].index
	local year = invitesToAccept[1].year
	local month = invitesToAccept[1].month
	C_Calendar.SetAbsMonth(month,year)
	local event = C_Calendar.GetDayEvent(0,d,i)
	local openEventIndex = C_Calendar.GetEventIndex()
	local openEventInfo = C_Calendar.GetEventInfo()

	if AIA:InviteStatus(event) == 'Replied' or AIA:InviteStatus(event) == false then
		table.remove(invitesToAccept,1)
		return
	end

	if openEventIndex then -- We have an event open
		if openEventIndex.offsetMonths == 0 and openEventIndex.monthDay == d and openEventIndex.eventIndex == i then
			-- We have an event open, and the data is all correct.
		else
			-- We have an event open, but it's not the correct one.
			C_Calendar.OpenEvent(0,d,i)
			return
		end
	else
		-- We don't have an event open so open one.
		C_Calendar.OpenEvent(0,d,i)
		return
	end

	if event and openEventInfo and AIA:InviteStatus(openEventInfo) == true then
		if event.inviteType == 1 then
			C_Calendar.EventAvailable()
		elseif event.inviteType == 2 then
			C_Calendar.EventSignUp()
		end
	end
end

function LDB:OnTooltipShow()
	self:AddLine('|cFFFF2C5AAIA|r')
	self:AddLine(' ')
	if eventCount == 1 then
		self:AddLine(string.format('|cFFFF2C5A%d|r |cFFFFFFFF'..L['Pending invite']..'|r', eventCount))
	elseif eventCount > 0 then
		self:AddLine(string.format('|cFFFF2C5A%d|r |cFFFFFFFF'..L['Pending invites']..'|r', eventCount))
	end
	if sessionAccepted == 1 then
		self:AddLine(string.format('|cFFFFBE19%d|r |cFFFFFFFF'..L['invite accepted this session']..'|r', sessionAccepted))
	elseif sessionAccepted > 0 then
		self:AddLine(string.format('|cFFFFBE19%d|r |cFFFFFFFF'..L['invites accepted this session']..'|r', sessionAccepted))
	else
		self:AddLine('|cFFFFFFFF'..L['No new invites accepted this session']..'|r')
	end
end

function LDB:OnClick(button)
	local click = {
		['LeftButton'] = AIA.db.profile.LDB.LeftClick,
		['RightButton'] = AIA.db.profile.LDB.RightClick,
		['MiddleButton'] = AIA.db.profile.LDB.MiddleClick,
	}
	if click[button] == 1 then
		AIA:CheckAgain()
	elseif click[button] == 2 then
		GameTimeFrame:Click()
	elseif click[button] == 3 then
		AIA:ChatCommand('Open')
	else return
	end
end

function LDB:OnEnter()
	GameTooltip:SetOwner(self, 'ANCHOR_NONE')
	GameTooltip:SetPoint('TOPLEFT', self, 'BOTTOMLEFT')
	GameTooltip:ClearLines()
	GameTooltip:Show()
end

function LDB:OnLeave()
	GameTooltip:Hide()
end

function AIA:LDBDisplayState()
	if AIA.db.profile.Minimap.hide then
		LibStub('LibDBIcon-1.0'):Hide('AIA')
	else
		LibStub('LibDBIcon-1.0'):Show('AIA')
	end
end

function AIA:LDBLockState()
	if AIA.db.profile.Minimap.lock then
		LibStub('LibDBIcon-1.0'):Lock('AIA')
	else
		LibStub('LibDBIcon-1.0'):Unlock('AIA')
	end
end

local function CalendarClosed()
	invitesAccepted = 0
	AIA:Enable()
	AIA:CreateCalendarList()
	AIA:AcceptInvite()
	CalendarFrame:SetScript('OnHide',nil)
end

function AIA:CheckAgain()
	if CalendarFrame then
		if CalendarFrame:IsVisible() == true then
			print('|cFFFF2C5AAIA: |r'..L['Will check again after you close the calendar.'])
			CalendarFrame:SetScript('OnHide',CalendarClosed)
			return
		end
	end
	print('|cFFFF2C5AAIA: |r'..L['Checking again!'])
	invitesAccepted = 0
	AIA:Enable()
	AIA:CreateCalendarList()
	AIA:AcceptInvite()
end

function AIA:OnEnable()
	AIA.ListMonitor = AIA.ListMonitor or CreateFrame('Frame', 'AIA_ListMonitor')
	AIA.ListMonitor:RegisterEvent('PLAYER_ENTERING_WORLD')
	AIA.ListMonitor:RegisterEvent('CALENDAR_UPDATE_EVENT_LIST')
	AIA.ListMonitor:RegisterEvent('CALENDAR_UPDATE_INVITE_LIST')
	AIA.ListMonitor:RegisterEvent('CALENDAR_ACTION_PENDING')
	AIA.ListMonitor:SetScript('OnEvent', AIA.CreateCalendarList)
	AIA.Accepter = AIA.Accepter or CreateFrame('Frame', 'AIA_Accepter')
end

function AIA:OnDisable()
	AIA.ListMonitor:UnregisterAllEvents()
	AIA.ListMonitor:SetScript('OnEvent', nil)
	AIA.Accepter:UnregisterAllEvents()
	AIA.Accepter:SetScript('OnUpdate', nil)
	if AIA.db.profile.NotifyDisable then
		if (AIA.db.profile.NotifyOnlyAccepted and invitesAccepted > 0) or not AIA.db.profile.NotifyOnlyAccepted then
			print('|cFFFF2C5AAIA: |r'..L['No more invites to accept. Shutting down.'])
		end
	end
end

Init.lua
local AIA = LibStub('AceAddon-3.0'):NewAddon('AIA', 'AceConsole-3.0', 'AceEvent-3.0')
local L = LibStub('AceLocale-3.0'):GetLocale('AIA')
local AceConfigDialog = LibStub('AceConfigDialog-3.0')
local LDB = LibStub('LibDataBroker-1.1'):NewDataObject('AIA', {type = 'data source', text = 'AIA'})

function AIA:TempOptions()
	local Options = {
		type = 'group',
		name = function(info)
			return 'AIA - |cFFFF2C5A'..L['Accept Invites Automatically'].. '|r r|cFFFF2C5A' .. string.match(GetAddOnMetadata('AIA','Version'),'%d+') ..'|r'
		end,
		order = 0,
		args = {
				Open = {
					name = L['Open Configuration Panel'],
					type = 'execute',
					order = 0,
					func = function()
						if not InCombatLockdown() then
							-- HideUIPanel protected in 8.2 Prevent UI taint.
							HideUIPanel(InterfaceOptionsFrame)
							HideUIPanel(GameMenuFrame)
							AIA:ChatCommand('Open')
						end
					end,
				},
			},
		}
	LibStub('AceConfig-3.0'):RegisterOptionsTable('AIA_Blizz', Options) -- Register Options
	self.optionsFrame = AceConfigDialog:AddToBlizOptions('AIA_Blizz', 'AIA')
end

function AIA:ChatCommand(input)
	if string.match(string.lower(input),'check') then
		AIA:CheckAgain()
	else
		if not InCombatLockdown() then
			self:EnableModule('Options')
			if AceConfigDialog.OpenFrames['AIA'] then
				AceConfigDialog:Close('AIA')
			else
				AceConfigDialog:Open('AIA')
			end
		else
			print('|cFFFF2C5AAIA: |r'..L['Cannot configure while in combat.'])
		end
	end
end

function AIA:OnInitialize()
	self.db = LibStub('AceDB-3.0'):New('AIADB', AIA.Defaults) -- Setup Saved Variables
	self:RegisterChatCommand('AIA', 'ChatCommand') -- Register /AIA command

	-- Profile Management
	self.db.RegisterCallback(self, 'OnProfileChanged', 'RefreshConfig')
	self.db.RegisterCallback(self, 'OnProfileCopied', 'RefreshConfig')
	self.db.RegisterCallback(self, 'OnProfileReset', 'RefreshConfig')

	-- Add Button to Open AIA's Options panel.
	AIA:TempOptions()

	LDB.text = AIA.db.profile.LDB.DisplayName
	LDB.icon = 'Interface\\Addons\\AIA\\Media\\AIA-Icon.tga'
	LibStub('LibDBIcon-1.0'):Register('AIA', LDB, AIA.db.profile.Minimap)
end

function AIA:RefreshConfig()
	AIA.db.profile = self.db.profile
	AIA:LDBDisplayState()
	AIA:LDBLockState()
end

Utilities.lua
local AIA = LibStub('AceAddon-3.0'):GetAddon('AIA')

function AIA:IsLeapYear(year)
	return year % 4 == 0 and (year % 100 ~= 0 or year % 400 == 0)
end

function AIA:DaysInMonth(month,year)
	return month == 2 and AIA:IsLeapYear(year) and 29 or ('\31\28\31\30\31\30\31\31\30\31\30\31'):byte(month)
end

function AIA:FindEventCreator(event)
	-- For some events invitedBy is an empty string, in these cases creator exists so check for either and return that name.
	if event.creator then
		if string.len(event.creator) > 0 then
			return event.creator
		end
	elseif event.invitedBy then
		if string.len(event.invitedBy) > 0 then
			return event.invitedBy
		end
	else
		return
	end
end

function AIA:SplitString(String)
	-- Split comma delimited strings into table entries.
	local Names = {}
	for word in string.gmatch(String,'[^,%s]+') do
		table.insert(Names,string.lower(word))
	end
	return Names
end

function AIA:StringFilter(List, Compare)
	-- Compare user entered name or names to whoever the event creator/inviter was.
	local Names = AIA:SplitString(List)
	if not string.match(List,'%a') then return true end -- If the user didn't enter any letters, there can't be any names, so just ignore it even if there is something else there.
	if not string.match(Compare,'%a') then return true end -- If there was no supplied Compare values, something funky happened, deal with that elsewhere.
	for i = 1,#Names do
		if Names[i] == string.lower(Compare) then return true end
	end
	return false
end

function AIA:EventFilter(event)
	for i = 0,4 do
		if AIA.db.profile.Filter.Type[i] and event.eventType == i then return true end -- eventType 0 == Raid, 1 == Dungeon, 2 == PvP, 3 == Meeting, 4 == Other
	end
	return false
end

function AIA:AddZero(number)
	-- Event date info returns dates without leading zeros, to easily compare dates, this adds them back in if necessary.
	local string = tostring(number)
	if tonumber(string) < 10 then
		string = '0'.. string
	end
	return string
end

function AIA:DateConversion(date)
	-- Takes the table of information for the date and formats it as one number for easy comparison to the current date.
	local year = date.year
	local month = date.month
	local monthDay = date.monthDay
	local hour = date.hour
	local minute = date.minute

	month = AIA:AddZero(month)
	monthDay = AIA:AddZero(monthDay)
	hour = AIA:AddZero(hour)
	minute = AIA:AddZero(minute)

	return tonumber(year..month..monthDay..hour..minute)
end

function AIA:CheckFilteredEventCount(table)
	local count = 0
	for k,v in pairs(table) do
		if v == true then
			count = count + 1
		end
	end
	return count
end

Options.lua
local AIA = LibStub('AceAddon-3.0'):GetAddon('AIA')
local L = LibStub('AceLocale-3.0'):GetLocale('AIA')
local AIA_Options = AIA:NewModule('Options')

local Options = {
	type = "group",
	name = function(info)
		return "AIA - |cFFFF2C5A"..L["Accept Invites Automatically"].. "|r r|cFFFF2C5A" .. string.match(GetAddOnMetadata("AIA","Version"),"%d+") .."|r"
	end,
	order = 0,
	childGroups = "tab",
	args = {
		Filters = {
			name = L["Filtering"],
			type = "group",
			order = 1,
			args = {
				Status = {
					name = L["Filter by status:"],
					type = "group",
					order = 0,
					inline = true,
					args = {
						Tentative = {
							name = "|cFFFF8019"..L["Tentative"].."|r",
							desc = L["Override Tentative status. Accepts calender invites where you have replied as tentative."],
							descStyle = "inline",
							width = "full",
							order = 10,
							type = "toggle",
							get = function(info)
								return AIA.db.profile.Types.Tentative
							end,
							set = function(info, value)
								AIA.db.profile.Types.Tentative = value
							end,
						},
						Delined = {
							name = "|CFFFF0000"..L["Declined"].."|r",
							desc = L["Override Declined status. Accepts calender invites where you have replied as declined."],
							descStyle = "inline",
							width = "full",
							order = 10,
							type = "toggle",
							get = function(info)
								return AIA.db.profile.Types.Declined
							end,
							set = function(info, value)
								AIA.db.profile.Types.Declined = value
							end,
						},
						Invited = {
							name = "|cFFFFBE19"..L["Invited"].."|r",
							desc = L["Accepts calendar invites where you haven't replied."],
							descStyle = "inline",
							width = "full",
							order = 10,
							type = "toggle",
							get = function(info)
								return AIA.db.profile.Types.Invited
							end,
							set = function(info, value)
								AIA.db.profile.Types.Invited = value
							end,
						},
						SignUp = {
							name = "|cFF91be0f"..L["Not Signed Up"].."|r",
							desc = L["Sign Up to guild or community events that you are not signed up to or have not been specifically invited to."],
							descStyle = "inline",
							width = "full",
							order = 30,
							type = "toggle",
							get = function(info)
								return AIA.db.profile.Types.SignUp
							end,
							set = function(info, value)
								AIA.db.profile.Types.SignUp = value
							end,
						},
					},
				},
				Names = {
					name = L["Filter by event creator/inviter name:"],
					type = "group",
					order = 1,
					inline = true,
					args = {
						FilterName = {
							name = function()
								if AIA.db.profile.Types.Invited == false and AIA.db.profile.Types.Declined == false and AIA.db.profile.Types.Tentative == false then
									return L["Only accept invites sent by:"]
								else
									return "|cFFFFCC00"..L["Only accept invites sent by:"].."|r"
								end
							end,
							desc = L["Only accept event invites if they were sent by this player."].."\n"..L["Leave blank to allow all."].."\n"..L["You can enter multiple names by separating each one with a comma."],
							order = 20,
							type = "input",
							width = 1.5,
							disabled = function()
								if AIA.db.profile.Types.Invited == false and AIA.db.profile.Types.Declined == false and AIA.db.profile.Types.Tentative == false then
									return true
								end
							end,
							get = function(info)
								return AIA.db.profile.Filter.Name
							end,
							set = function(info, value)
								AIA.db.profile.Filter.Name = value
							end,
						},
						Filter_SignUp_Name = {
							name = function()
								if AIA.db.profile.Types.SignUp == false then
									return L["Only sign up to events created by:"]
								else
									return "|cFF91be0f"..L["Only sign up to events created by:"].."|r"
								end
							end,
							desc = L["Only sign up to events if they were created by this player."].."\n"..L["Leave blank to allow all."].."\n"..L["You can enter multiple names by separating each one with a comma."],
							order = 33,
							type = "input",
							width = 1.5,
							disabled = function()
								if AIA.db.profile.Types.SignUp == false then return true end
							end,
							get = function(info)
								return AIA.db.profile.Filter.SignUp.Name
							end,
							set = function(info, value)
								AIA.db.profile.Filter.SignUp.Name = value
							end,
						},
					},
				},
				Titles = {
					name = L["Filter by event title:"],
					type = "group",
					order = 2,
					inline = true,
					args = {
						FilterTitle = {
							name = function()
								if AIA.db.profile.Types.Invited == false and AIA.db.profile.Types.Declined == false and AIA.db.profile.Types.Tentative == false then
									return L["Only accept invites with title:"]
								else
									return "|cFFFFCC00"..L["Only accept invites with title:"].."|r"
								end
							end,
							desc = L["Only accept event invites if the title of the event matches this."].."\n"..L["Leave blank to allow all."].."\n"..L["You can enter multiple events by separating each one with a comma."],
							order = 22,
							type = "input",
							width = 1.5,
							disabled = function()
								if AIA.db.profile.Types.Invited == false and AIA.db.profile.Types.Declined == false and AIA.db.profile.Types.Tentative == false then
									return true
								end
							end,
							get = function(info)
								return AIA.db.profile.Filter.Title
							end,
							set = function(info, value)
								AIA.db.profile.Filter.Title = value
							end,
						},
						Filter_SignUp_Title = {
							name = function()
								if AIA.db.profile.Types.SignUp == false then
									return L["Only sign up to events with title:"]
								else
									return "|cFF91be0f"..L["Only sign up to events with title:"].."|r"
								end
							end,
							desc = L["Only sign up to events if the title of the event matches this."].."\n"..L["Leave blank to allow all."].."\n"..L["You can enter multiple events by separating each one with a comma."],
							order = 35,
							type = "input",
							width = 1.5,
							disabled = function()
								if AIA.db.profile.Types.SignUp == false then return true end
							end,
							get = function(info)
								return AIA.db.profile.Filter.SignUp.Title
							end,
							set = function(info, value)
								AIA.db.profile.Filter.SignUp.Title = value
							end,
						},
					},
				},
				FilterType = {
					name = "|cFFFFCC00"..L["Filter by event type:"].."|r",
					desc = L["Only accept invites if the event type matches this."],
					order = 40,
					type = "multiselect",
					values = {
						[0] = L["Raid"],
						[1] = L["Dungeon"],
						[2] = L["PvP"],
						[3] = L["Meeting"],
						[4] = L["Other"]
					},
					get = function(info,key)
						return AIA.db.profile.Filter.Type[key]
					end,
					set = function(info, key, value)
						AIA.db.profile.Filter.Type[key] = value
					end,
				},
			},
		},
		Notifications = {
			name = L["Notifications"],
			type = "group",
			order = 0,
			args = {
				DisableAfterRunning = {
					name = "|cFF91be0f"..L["Disable After Running"].."|r",
					desc = L["Sets AIA to disable itself when it has accepted all the invites it is allowed to."].."\n"..L["If left unchecked, AIA will continue to scan and accept any new calendar events while you are playing."],
					descStyle = "inline",
					width = "full",
					order = 0,
					type = "toggle",
					get = function(info)
						return AIA.db.profile.DisableAfterRunning
					end,
					set = function(info, value)
						AIA.db.profile.DisableAfterRunning = value
					end,
				},
				NotifyDisable = {
					name = function()
						if AIA.db.profile.DisableAfterRunning == false then
							return L["Notify when disabling"]
						else
							return "|cFF00b2fa"..L["Notify when disabling"].."|r"
						end
					end,
					desc = L["Sets if AIA should send a message to the chat window when it is disabled."],
					descStyle = "inline",
					width = "full",
					order = 1,
					type = "toggle",
					disabled = function()
						if AIA.db.profile.DisableAfterRunning == false then return true end
					end,
					get = function(info)
						return AIA.db.profile.NotifyDisable
					end,
					set = function(info, value)
						AIA.db.profile.NotifyDisable = value
					end,
				},
				Spacer1 = {
					name = " ",
					type = "description",
					width = "normal",
					order = 9,
				},
				NotifyFinish = {
					name = "|cFF00b2fa"..L["Notify when finished"].."|r",
					desc = L["Sets if AIA should send a message to the chat window when it has finished."],
					descStyle = "inline",
					width = "full",
					order = 10,
					type = "toggle",
					get = function(info)
						return AIA.db.profile.NotifyFinish
					end,
					set = function(info, value)
						AIA.db.profile.NotifyFinish = value
					end,
				},
				OnlyNotifyIfAccepted = {
					name = function()
						if AIA.db.profile.NotifyFinish == false then
							return L["Notify only accepted invites"]
						else
							return "|cFF00b2fa"..L["Notify only accepted invites"].."|r"
						end
					end,
					desc = L["AIA will only notify when finished if there were any invites to accept."],
					descStyle = "inline",
					width = "full",
					order = 11,
					type = "toggle",
					disabled = function()
						if AIA.db.profile.NotifyFinish == false then return true end
					end,
					get = function(info)
						return AIA.db.profile.NotifyOnlyAccepted
					end,
					set = function(info, value)
						AIA.db.profile.NotifyOnlyAccepted = value
					end,
				},
				Spacer2 = {
					name = " ",
					type = "description",
					width = "normal",
					order = 19,
				},
				WarnIgnoredEvents = {
					name = "|cFFFF2C5A"..L["Warn about ignored events"].."|r",
					desc = L["Sets if AIA will send a message to the chat window if it ignores an event that does not meet the settings below. As a notification that you have unanswered events to deal with yourself."],
					descStyle = "inline",
					width = "full",
					order = 20,
					type = "toggle",
					get = function(info)
						return AIA.db.profile.WarnIgnoredEvents
					end,
					set = function(info, value)
						AIA.db.profile.WarnIgnoredEvents = value
					end,
				},
			},
		},
		LDB = {
			name = L["Minimap"],
			type = "group",
			order = 60,
			args = {
				Display = {
					name = function()
						if AIA.db.profile.Minimap.hide then
							return "|cFFd83636"..L["Hide Minimap Icon"].."|r"
						else
							return "|cFF82d836"..L["Hide Minimap Icon"].."|r"
						end
					end,
					desc = L["Even if the minimap button is disabled, AIA still provides an LDB plugin if you want it to be displayed with addons such as StatBlockCore or TitanPanel."],
					width = 3,
					order = 0,
					type = "toggle",
					get = function(info)
						return AIA.db.profile.Minimap.hide
					end,
					set = function(info, value)
						AIA.db.profile.Minimap.hide = value
						AIA:LDBDisplayState()
					end,
				},
				Lock = {
					name = function()
						if AIA.db.profile.Minimap.lock then
							return "|cFFd83636"..L["Lock Minimap Icon"].."|r"
						else
							return "|cFF82d836"..L["Lock Minimap Icon"].."|r"
						end
					end,
					width = 3,
					order = 1,
					type = "toggle",
					get = function(info)
						return AIA.db.profile.Minimap.lock
					end,
					set = function(info, value)
						AIA.db.profile.Minimap.lock = value
						AIA:LDBLockState()
					end,
				},
				MouseInteraction = {
					name = L["Mouse Button Interaction"],
					type = "group",
					inline = true,
					order = 30,
					args = {
						LeftClick = {
							name = "|cFFFFCC00"..L["Left Click functionality"].."|r",
							order = 40,
							type = "select",
							values = {
								[0] = L["Do Nothing"],
								[1] = L["Check Again"],
								[2] = L["Open Calendar"],
								[3] = L["Open Options"],
							},
							get = function(info,key)
								return AIA.db.profile.LDB.LeftClick
							end,
							set = function(info, key, value)
								AIA.db.profile.LDB.LeftClick = value
							end,
						},
						RightClick = {
							name = "|cFFFFCC00"..L["Right Click functionality"].."|r",
							order = 41,
							type = "select",
							values = {
								[0] = L["Do Nothing"],
								[1] = L["Check Again"],
								[2] = L["Open Calendar"],
								[3] = L["Open Options"],
							},
							get = function(info,key)
								return AIA.db.profile.LDB.RightClick
							end,
							set = function(info, key, value)
								AIA.db.profile.LDB.RightClick = value
							end,
						},
						MiddleClick = {
							name = "|cFFFFCC00"..L["Middle Click functionality"].."|r",
							order = 42,
							type = "select",
							values = {
								[0] = L["Do Nothing"],
								[1] = L["Check Again"],
								[2] = L["Open Calendar"],
								[3] = L["Open Options"],
							},
							get = function(info,key)
								return AIA.db.profile.LDB.MiddleClick
							end,
							set = function(info, key, value)
								AIA.db.profile.LDB.MiddleClick = value
							end,
						},
					},
				},
			},
		},
		CheckGroup = {
			name = " ",
			type = "group",
			inline = true,
			args = {
				CheckAgainSpacer = {
					name = " ",
					type = "description",
					width = "normal",
				},
				CheckAgain = {
					name = "|cFF40c840"..L["Check Again"].."|r",
					desc = L["Makes AIA Check if there is anything to accept now."],
					type = "execute",
					func = function()
						print("|cFFFF2C5AAIA: |r"..L["Checking again!"])
						AIA:Enable()
						AIA:CreateCalendarList()
						AIA:AcceptInvite()
					end,
				},
			},
		},

	},
}

function AIA_Options:OnInitialize()
	self:SetEnabledState(false)
end

function AIA_Options:OnEnable()
	self.db = AIA.db -- Setup Saved Variables

	-- Add Options
	LibStub('AceConfigRegistry-3.0'):RegisterOptionsTable('AIA', Options)
	local Profiles = LibStub('AceDBOptions-3.0'):GetOptionsTable(self.db)
	LibStub('AceConfigDialog-3.0'):SetDefaultSize('AIA',590,715)
	Options.args.profiles = Profiles
	Options.args.profiles.order = 99

	-- Profile Management
	self.db.RegisterCallback(self, 'OnProfileChanged', 'RefreshConfig')
	self.db.RegisterCallback(self, 'OnProfileCopied', 'RefreshConfig')
	self.db.RegisterCallback(self, 'OnProfileReset', 'RefreshConfig')
end

function AIA_Options:RefreshConfig()
	AIA.db.profile = self.db.profile
end